File: routes/secureLicenses.js
const express = require('express');
const crypto = require('crypto');
const SecureLicense = require('../models/SecureLicense');
const MilitaryGradeSecurity = require('../services/militaryGradeSecurity');
const { auth, adminOnly } = require('../middleware/auth');
const rateLimit = require('express-rate-limit');

const router = express.Router();
const security = new MilitaryGradeSecurity();

// Rate limiting for license validation
const validationLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100, // limit each IP to 100 requests per windowMs
  message: {
    error: 'Too many validation requests',
    message: 'Please try again later'
  },
  standardHeaders: true,
  legacyHeaders: false
});

// Get all secure licenses (admin only)
router.get('/', auth, adminOnly, async (req, res) => {
  try {
    const { 
      page = 1, 
      limit = 10, 
      status, 
      licenseType, 
      securityLevel,
      riskScore 
    } = req.query;
    
    const query = {};
    
    if (status) query.status = status;
    if (licenseType) query.licenseType = licenseType;
    if (securityLevel) query.securityLevel = securityLevel;
    if (riskScore) query['securityMetrics.riskScore'] = { $gte: parseInt(riskScore) };

    const licenses = await SecureLicense.find(query)
      .populate('createdBy', 'email')
      .sort({ createdAt: -1 })
      .limit(limit * 1)
      .skip((page - 1) * limit)
      .select('-encryptionKeys -integrityChecksums'); // Exclude sensitive data

    const total = await SecureLicense.countDocuments(query);

    res.json({
      licenses,
      totalPages: Math.ceil(total / limit),
      currentPage: page,
      total
    });
  } catch (error) {
    console.error('Get secure licenses error:', error);
    res.status(500).json({ error: 'Failed to fetch licenses' });
  }
});

// Get secure license by ID (admin only)
router.get('/:id', auth, adminOnly, async (req, res) => {
  try {
    const license = await SecureLicense.findById(req.params.id)
      .populate('createdBy', 'email')
      .select('-encryptionKeys'); // Exclude encryption keys

    if (!license) {
      return res.status(404).json({ error: 'License not found' });
    }

    res.json(license);
  } catch (error) {
    console.error('Get secure license error:', error);
    res.status(500).json({ error: 'Failed to fetch license' });
  }
});

// Create new secure license (admin only)
router.post('/', auth, adminOnly, async (req, res) => {
  try {
    const {
      clientName,
      clientEmail,
      productName = 'Torro Platform',
      version = '1.0.0',
      licenseType = 'trial',
      securityLevel = 'basic',
      expiryDays = 30,
      maxUsers = 1,
      maxConnections = 10,
      features = {},
      hardwareBinding = false,
      selfDestruction = false,
      networkValidation = false,
      allowedIPs = [],
      allowedCountries = [],
      notes
    } = req.body;

    // Generate military-grade license key
    const licenseKey = security.generateSecureLicenseKey();
    
    // Generate truly unique client ID with multiple entropy sources
    const timestamp = Date.now();
    const randomPart1 = crypto.randomBytes(4).toString('hex').toUpperCase();
    const randomPart2 = Math.random().toString(36).substr(2, 9).toUpperCase();
    const processId = process.pid.toString(36).toUpperCase();
    const clientId = `TORRO-${timestamp}-${randomPart1}-${randomPart2}-${processId}`;

    // Calculate expiry date
    const expiryDate = new Date();
    expiryDate.setDate(expiryDate.getDate() + expiryDays);

    // Generate hardware fingerprint if required
    let hardwareFingerprint = null;
    if (securityLevel === 'military' || hardwareBinding) {
      hardwareFingerprint = security.generateHardwareFingerprint();
    }

    // Create license data for encryption
    const licenseData = {
      licenseKey,
      clientId,
      clientName,
      clientEmail,
      productName,
      version,
      licenseType,
      expiryDate,
      maxUsers,
      maxConnections,
      features,
      securityLevel
    };

    // Encrypt license data if military-grade
    let encryptedData = null;
    if (securityLevel === 'military') {
      encryptedData = security.encryptLicenseData(licenseData, hardwareFingerprint);
    }

    // Check for existing license with same client ID (extra safety)
    const existingLicense = await SecureLicense.findOne({ clientId });
    if (existingLicense) {
      // Regenerate client ID if collision detected
      const newTimestamp = Date.now() + Math.floor(Math.random() * 1000);
      const newRandomPart1 = crypto.randomBytes(4).toString('hex').toUpperCase();
      const newRandomPart2 = Math.random().toString(36).substr(2, 9).toUpperCase();
      const newProcessId = process.pid.toString(36).toUpperCase();
      clientId = `TORRO-${newTimestamp}-${newRandomPart1}-${newRandomPart2}-${newProcessId}`;
    }

    // Create secure license
    const secureLicense = new SecureLicense({
      licenseKey,
      clientId,
      clientName,
      clientEmail,
      productName,
      version,
      licenseType,
      securityLevel,
      expiryDate,
      maxUsers,
      maxConnections,
      features: new Map(Object.entries(features)),
      hardwareFingerprint,
      encryptionKeys: encryptedData ? {
        layer1Key: encryptedData.encrypted,
        layer2Key: encryptedData.checksums.sha256,
        layer3PrivateKey: encryptedData.privateKey,
        keyRotationTimestamp: new Date()
      } : undefined,
      integrityChecksums: {
        sha256: security.generateChecksums(JSON.stringify(licenseData)).sha256,
        sha512: security.generateChecksums(JSON.stringify(licenseData)).sha512,
        custom: security.generateCustomChecksum(JSON.stringify(licenseData))
      },
      networkValidation: {
        enabled: networkValidation,
        allowedIPs,
        allowedCountries,
        maxRequestsPerHour: 1000
      },
      selfDestruction: {
        enabled: selfDestruction,
        gracePeriod: 300000, // 5 minutes
        destructionTriggers: ['expiry', 'tampering']
      },
      notes,
      createdBy: req.user.userId
    });

    await secureLicense.save();

    // Add audit trail
    secureLicense.auditTrail.push({
      action: 'created',
      performedBy: req.user.userId,
      details: new Map([
        ['securityLevel', securityLevel],
        ['hardwareBinding', hardwareBinding.toString()],
        ['selfDestruction', selfDestruction.toString()]
      ])
    });

    await secureLicense.save();

    res.status(201).json({
      message: 'Secure license created successfully',
      license: secureLicense.getClientInfo()
    });
  } catch (error) {
    console.error('Create secure license error:', error);
    if (error.code === 11000) {
      res.status(400).json({ error: 'License key or client ID already exists' });
    } else {
      res.status(500).json({ error: 'Failed to create secure license' });
    }
  }
});

// Validate secure license (public endpoint with rate limiting)
router.post('/validate', validationLimiter, async (req, res) => {
  try {
    const { licenseKey, clientId, hardwareFingerprint } = req.body;

    if (!licenseKey || !clientId) {
      return res.status(400).json({ 
        error: 'Missing required fields',
        message: 'License key and client ID are required'
      });
    }

    // Find license
    const license = await SecureLicense.findByLicenseKey(licenseKey);
    
    if (!license) {
      return res.status(404).json({ 
        error: 'Invalid license',
        valid: false,
        code: 'LICENSE_NOT_FOUND'
      });
    }

    // Verify client ID
    if (license.clientId !== clientId) {
      license.addValidationRecord({
        success: false,
        ipAddress: req.ip,
        userAgent: req.get('User-Agent'),
        securityViolations: ['client_id_mismatch']
      });
      await license.save();

      return res.status(403).json({
        error: 'Client ID mismatch',
        valid: false,
        code: 'CLIENT_ID_MISMATCH'
      });
    }

    // Check if license is valid
    if (!license.isValid()) {
      return res.status(403).json({
        error: 'License is not valid',
        valid: false,
        code: license.status === 'expired' ? 'LICENSE_EXPIRED' : 'LICENSE_INVALID',
        expiryDate: license.expiryDate,
        status: license.status
      });
    }

    // Hardware fingerprint validation for military-grade licenses
    if (license.securityLevel === 'military') {
      if (!hardwareFingerprint) {
        return res.status(400).json({
          error: 'Hardware fingerprint required',
          valid: false,
          code: 'HARDWARE_FINGERPRINT_REQUIRED'
        });
      }

      if (!license.validateHardwareFingerprint(hardwareFingerprint)) {
        license.addValidationRecord({
          success: false,
          ipAddress: req.ip,
          userAgent: req.get('User-Agent'),
          securityViolations: ['hardware_mismatch']
        });
        await license.save();

        return res.status(403).json({
          error: 'Hardware fingerprint mismatch',
          valid: false,
          code: 'HARDWARE_MISMATCH'
        });
      }
    }

    // Network validation
    if (license.networkValidation.enabled) {
      const clientIP = req.ip;
      
      // Check IP whitelist
      if (license.networkValidation.allowedIPs.length > 0) {
        if (!license.networkValidation.allowedIPs.includes(clientIP)) {
          license.addValidationRecord({
            success: false,
            ipAddress: clientIP,
            userAgent: req.get('User-Agent'),
            securityViolations: ['ip_not_allowed']
          });
          await license.save();

          return res.status(403).json({
            error: 'IP address not allowed',
            valid: false,
            code: 'IP_NOT_ALLOWED'
          });
        }
      }

      // Check rate limiting
      const now = new Date();
      const oneHourAgo = new Date(now.getTime() - 60 * 60 * 1000);
      
      if (license.networkValidation.lastRequestReset < oneHourAgo) {
        license.networkValidation.currentRequests = 0;
        license.networkValidation.lastRequestReset = now;
      }

      if (license.networkValidation.currentRequests >= license.networkValidation.maxRequestsPerHour) {
        return res.status(429).json({
          error: 'Rate limit exceeded',
          valid: false,
          code: 'RATE_LIMIT_EXCEEDED'
        });
      }

      license.networkValidation.currentRequests++;
    }

    // Record successful validation
    license.addValidationRecord({
      success: true,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      hardwareMatch: license.securityLevel === 'military' ? true : undefined,
      responseTime: Date.now() - req.startTime
    });

    await license.save();

    // Return license info
    res.json({
      valid: true,
      license: license.getClientInfo(),
      securityLevel: license.securityLevel,
      riskScore: license.securityMetrics.riskScore
    });

  } catch (error) {
    console.error('Validate secure license error:', error);
    res.status(500).json({ 
      error: 'License validation failed',
      message: 'Internal server error'
    });
  }
});

// Update secure license (admin only)
router.put('/:id', auth, adminOnly, async (req, res) => {
  try {
    const updates = req.body;
    
    // Remove sensitive fields that shouldn't be updated directly
    delete updates.licenseKey;
    delete updates.clientId;
    delete updates.encryptionKeys;
    delete updates.integrityChecksums;
    delete updates.validationHistory;

    const license = await SecureLicense.findByIdAndUpdate(
      req.params.id,
      updates,
      { new: true, runValidators: true }
    );

    if (!license) {
      return res.status(404).json({ error: 'License not found' });
    }

    // Add audit trail
    license.auditTrail.push({
      action: 'updated',
      performedBy: req.user.userId,
      details: new Map(Object.entries(updates))
    });

    await license.save();

    res.json({
      message: 'Secure license updated successfully',
      license: license.getClientInfo()
    });
  } catch (error) {
    console.error('Update secure license error:', error);
    res.status(500).json({ error: 'Failed to update secure license' });
  }
});

// Delete secure license (admin only)
router.delete('/:id', auth, adminOnly, async (req, res) => {
  try {
    const license = await SecureLicense.findByIdAndDelete(req.params.id);
    
    if (!license) {
      return res.status(404).json({ error: 'License not found' });
    }

    res.json({ message: 'Secure license deleted successfully' });
  } catch (error) {
    console.error('Delete secure license error:', error);
    res.status(500).json({ error: 'Failed to delete secure license' });
  }
});

// Get license statistics (admin only)
router.get('/stats/overview', auth, adminOnly, async (req, res) => {
  try {
    const total = await SecureLicense.countDocuments();
    const active = await SecureLicense.countDocuments({ status: 'active' });
    const expired = await SecureLicense.countDocuments({ status: 'expired' });
    const compromised = await SecureLicense.countDocuments({ status: 'compromised' });
    const military = await SecureLicense.countDocuments({ securityLevel: 'military' });
    const highRisk = await SecureLicense.countDocuments({ 'securityMetrics.riskScore': { $gte: 70 } });

    // Expiring soon (within 7 days)
    const sevenDaysFromNow = new Date();
    sevenDaysFromNow.setDate(sevenDaysFromNow.getDate() + 7);
    
    const expiringSoon = await SecureLicense.countDocuments({
      status: 'active',
      expiryDate: { $lte: sevenDaysFromNow, $gte: new Date() }
    });

    res.json({
      total,
      active,
      expired,
      compromised,
      military,
      highRisk,
      expiringSoon
    });
  } catch (error) {
    console.error('Get secure license stats error:', error);
    res.status(500).json({ error: 'Failed to fetch statistics' });
  }
});

// Get high-risk licenses (admin only)
router.get('/security/high-risk', auth, adminOnly, async (req, res) => {
  try {
    const highRiskLicenses = await SecureLicense.findHighRisk()
      .select('licenseKey clientName clientEmail securityMetrics.riskScore status')
      .sort({ 'securityMetrics.riskScore': -1 })
      .limit(50);

    res.json(highRiskLicenses);
  } catch (error) {
    console.error('Get high-risk licenses error:', error);
    res.status(500).json({ error: 'Failed to fetch high-risk licenses' });
  }
});

// Security violation endpoint (for client reporting)
router.post('/security/violation', validationLimiter, async (req, res) => {
  try {
    const { licenseKey, clientId, violationType, details } = req.body;

    if (!licenseKey || !clientId || !violationType) {
      return res.status(400).json({ 
        error: 'Missing required fields',
        message: 'License key, client ID, and violation type are required'
      });
    }

    const license = await SecureLicense.findByLicenseKey(licenseKey);
    
    if (!license || license.clientId !== clientId) {
      return res.status(404).json({ error: 'License not found' });
    }

    // Record security violation
    license.addValidationRecord({
      success: false,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
      securityViolations: [violationType]
    });

    // Add audit trail
    license.auditTrail.push({
      action: 'security_violation',
      details: new Map([
        ['violationType', violationType],
        ['details', JSON.stringify(details)]
      ])
    });

    await license.save();

    res.json({ message: 'Security violation recorded' });
  } catch (error) {
    console.error('Record security violation error:', error);
    res.status(500).json({ error: 'Failed to record security violation' });
  }
});

module.exports = router;

File: routes/secure_licenses.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from middleware.auth import auth_required, admin_required
from models import License, db
from services.license_service import validate_military_grade_license, calculate_risk_score
from datetime import datetime
import logging

secure_licenses_bp = Blueprint('secure_licenses', __name__)
logger = logging.getLogger(__name__)

@secure_licenses_bp.route('/validate', methods=['POST'])
def validate_secure_license():
    """Validate secure license with enhanced security checks"""
    try:
        data = request.get_json()
        license_key = data.get('licenseKey')
        client_id = data.get('clientId')
        hardware_fingerprint = data.get('hardwareFingerprint')
        timestamp = data.get('timestamp')
        user_agent = data.get('userAgent')
        
        if not license_key or not client_id:
            return jsonify({'error': 'License key and client ID are required'}), 400
        
        # Find license
        license = License.query.filter_by(license_key=license_key, client_id=client_id).first()
        
        if not license:
            return jsonify({
                'error': 'Invalid license',
                'valid': False
            }), 404
        
        # Update access tracking
        license.last_checked = datetime.utcnow()
        license.check_count += 1
        license.last_access_ip = request.remote_addr
        license.last_access_user_agent = request.headers.get('User-Agent', user_agent)
        
        # Check basic validity
        if not license.is_valid():
            return jsonify({
                'error': 'License has expired or is inactive',
                'valid': False,
                'expiryDate': license.expiry_date.isoformat(),
                'status': license.status
            }), 403
        
        # Enhanced security validation
        validation_attempt = {
            'license_key': license_key,
            'client_id': client_id,
            'hardware_fingerprint': hardware_fingerprint,
            'ip': request.remote_addr,
            'user_agent': request.headers.get('User-Agent', user_agent),
            'timestamp': timestamp or int(datetime.utcnow().timestamp()),
            'request_count': license.check_count
        }
        
        # Military-grade validation
        military_validation = validate_military_grade_license(license, validation_attempt)
        risk_score = calculate_risk_score(license, validation_attempt)
        
        if not military_validation.get('valid', True):
            db.session.commit()  # Save the violation
            return jsonify({
                'error': 'Enhanced security validation failed',
                'valid': False,
                'reason': military_validation.get('reason', 'Security violation detected'),
                'riskScore': military_validation.get('risk_score', risk_score)
            }), 403
        
        # Save updated license
        db.session.commit()
        
        # Prepare enhanced response
        response = {
            'valid': True,
            'license': license.get_client_info(show_encrypted_key=True),
            'riskScore': risk_score,
            'securityLevel': 'military',
            'militaryGrade': True,
            'hardwareMatch': military_validation.get('hardware_match', True),
            'integrityValid': military_validation.get('integrity_valid', True),
            'debuggerDetected': military_validation.get('debugger_detected', False),
            'securityFeatures': {
                'hardwareBinding': license.hardware_binding,
                'antiTampering': True,
                'antiDebugging': True,
                'selfDestruction': True,
                'integrityChecking': True,
                'riskScoring': True,
                'secureDeletion': True,
                'daemonSystem': True
            },
            'validationDetails': {
                'timestamp': datetime.utcnow().isoformat(),
                'ip': request.remote_addr,
                'userAgent': request.headers.get('User-Agent', user_agent),
                'checkCount': license.check_count
            }
        }
        
        return jsonify(response)
        
    except Exception as error:
        logger.error(f'Validate secure license error: {error}')
        return jsonify({'error': 'Secure license validation failed'}), 500

@secure_licenses_bp.route('/hardware-binding/<license_id>', methods=['POST'])
@admin_required
def update_hardware_binding(license_id):
    """Update hardware binding for a license (admin only)"""
    try:
        data = request.get_json()
        hardware_fingerprint = data.get('hardwareFingerprint')
        hardware_components = data.get('hardwareComponents', {})
        
        license = License.query.get(license_id)
        if not license:
            return jsonify({'error': 'License not found'}), 404
        
        # Update hardware binding
        license.hardware_fingerprint = hardware_fingerprint
        license.set_hardware_components(hardware_components)
        license.hardware_binding = True
        license.updated_at = datetime.utcnow()
        
        db.session.commit()
        
        return jsonify({
            'message': 'Hardware binding updated successfully',
            'license_id': license_id,
            'hardware_fingerprint': hardware_fingerprint
        })
        
    except Exception as error:
        logger.error(f'Update hardware binding error: {error}')
        db.session.rollback()
        return jsonify({'error': 'Failed to update hardware binding'}), 500

@secure_licenses_bp.route('/security-level/<license_id>', methods=['PUT'])
@admin_required
def update_security_level(license_id):
    """Update security level for a license (admin only)"""
    try:
        data = request.get_json()
        security_level = data.get('securityLevel')
        military_grade = data.get('militaryGrade', True)
        hardware_binding = data.get('hardwareBinding', True)
        
        if security_level not in ['basic', 'military']:
            return jsonify({'error': 'Invalid security level'}), 400
        
        license = License.query.get(license_id)
        if not license:
            return jsonify({'error': 'License not found'}), 404
        
        # Update security settings
        license.security_level = security_level
        license.military_grade = military_grade
        license.hardware_binding = hardware_binding
        license.updated_at = datetime.utcnow()
        
        db.session.commit()
        
        return jsonify({
            'message': 'Security level updated successfully',
            'license_id': license_id,
            'security_level': security_level,
            'military_grade': military_grade,
            'hardware_binding': hardware_binding
        })
        
    except Exception as error:
        logger.error(f'Update security level error: {error}')
        db.session.rollback()
        return jsonify({'error': 'Failed to update security level'}), 500

@secure_licenses_bp.route('/integrity-check/<license_id>', methods=['POST'])
@admin_required
def perform_integrity_check(license_id):
    """Perform integrity check on a license (admin only)"""
    try:
        license = License.query.get(license_id)
        if not license:
            return jsonify({'error': 'License not found'}), 404
        
        # Perform integrity validation
        validation_attempt = {
            'license_key': license.license_key,
            'client_id': license.client_id,
            'timestamp': int(datetime.utcnow().timestamp())
        }
        
        military_validation = validate_military_grade_license(license, validation_attempt)
        
        integrity_result = {
            'license_id': license_id,
            'integrity_valid': military_validation.get('integrity_valid', True),
            'hardware_match': military_validation.get('hardware_match', True),
            'debugger_detected': military_validation.get('debugger_detected', False),
            'risk_score': license.risk_score,
            'violations_count': len(license.get_security_violations()),
            'check_timestamp': datetime.utcnow().isoformat()
        }
        
        return jsonify(integrity_result)
        
    except Exception as error:
        logger.error(f'Integrity check error: {error}')
        return jsonify({'error': 'Failed to perform integrity check'}), 500

@secure_licenses_bp.route('/security-report/<license_id>', methods=['GET'])
@admin_required
def get_security_report(license_id):
    """Get comprehensive security report for a license (admin only)"""
    try:
        license = License.query.get(license_id)
        if not license:
            return jsonify({'error': 'License not found'}), 404
        
        # Gather security information
        security_violations = license.get_security_violations()
        hardware_components = license.get_hardware_components()
        integrity_checksums = license.get_integrity_checksums()
        encrypted_data = license.get_encrypted_data()
        
        security_report = {
            'license_id': license_id,
            'client_id': license.client_id,
            'client_name': license.client_name,
            'security_level': license.security_level,
            'military_grade': license.military_grade,
            'hardware_binding': license.hardware_binding,
            'hardware_fingerprint': license.hardware_fingerprint,
            'hardware_components': hardware_components,
            'integrity_checksums': integrity_checksums,
            'encrypted_data': encrypted_data,
            'risk_score': license.risk_score,
            'security_violations': security_violations,
            'access_statistics': {
                'check_count': license.check_count,
                'last_checked': license.last_checked.isoformat() if license.last_checked else None,
                'last_access_ip': license.last_access_ip,
                'last_access_user_agent': license.last_access_user_agent
            },
            'report_timestamp': datetime.utcnow().isoformat()
        }
        
        return jsonify(security_report)
        
    except Exception as error:
        logger.error(f'Get security report error: {error}')
        return jsonify({'error': 'Failed to generate security report'}), 500
